Generic Case Repeater

proteusvacuum opened the pull request:
## Product Description
Here's a big 🔨 

This allows you to create a case repeater based on UCR filters and expressions (with the open possibility to do the same thing with forms, but YAGNI).

This allows us to forward cases to any _existing_ API that has a well defined JSON spec without any custom dev work when app properties change, etc. The motivation for this was to integrate with [DIVOC](https://egovernments.github.io/DIVOC/developer-docs/api/admin-api.html#/vaccination-api.yaml). 

The expression must be a ~`dict`~ valid UCR expression that defines each property of the API. Thanks to the power of UCR expressions, this can also do related document lookups, etc. 

I believe that this can fulfill almost all of our custom API integration work, and pushes this down to end users. The experience of this will only get better as the tooling around UCR is developed.

The one major thing this is currently missing is how to handle a response (e.g. writing back to the case). However, this model could also be used in that scenario by creating a case update as long as the response is well formed JSON. 

I started writing this up as a spec, but realized it was a minimal amount of work to actually get it working, so opening this as a draft to get some feedback.

![Screenshot from 2021-11-15 17-00-25](https://user-images.githubusercontent.com/146896/141859936-a6aa0d25-30c8-4d04-95d3-8e51a2b11a13.png)


## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
- [x] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

proteusvacuum committed:
Reload json value if it is already the proper type

proteusvacuum committed:
Add expression repeater toggle

proteusvacuum committed:
Add Expression Repeater and Repeater Generator

proteusvacuum committed:
Add expression repeater views

czue commented on corehq/apps/userreports/ui/fields.py:
```diff
@@ -47,6 +47,9 @@ def prepare_value(self, value):
 
     def to_python(self, value):
+        if isinstance(value, self.expected_type):
+            return json.loads(json.dumps(value))
```
What problem does this solve? Or rather what previous assumption about values being passed in, no longer holds?

kaapstorm replied:
Do unit tests for this already exist? It would be cool to add a test that this change causes to pass.

czue commented on corehq/toggles.py:
```diff
@@ -2127,6 +2127,12 @@ def _commtrackify(domain_name, toggle_is_enabled):
+EXPRESSION_REPEATER = StaticToggle(
+    'expression_repeater',
+    'Integrate with generic APIs using UCR expressions',
+    TAG_SOLUTIONS_OPEN,
```
UCR itself is `TAG_SOLUTIONS_LIMITED`, and I think this feature has the possibility to do dangerous/non-performant things in the same fashion. Perhaps better to start more restrictive?

proteusvacuum replied:
Updated, although that UCR tag is going to change soon.

czue commented on corehq/motech/repeaters/expression/repeater_generators.py:
```diff
@@ -0,0 +1,18 @@
+    def get_payload(self, repeat_record, payload_doc, expression):
+        parsed_expression = ExpressionFactory.from_spec(expression, FactoryContext.empty())
+        result = parsed_expression(payload_doc, EvaluationContext(payload_doc))
+        return json.dumps(result, cls=DjangoJSONEncoder)
```
:heart: how easy this was.

czue commented on corehq/motech/repeaters/expression/repeaters.py:
```diff
@@ -0,0 +1,47 @@
+        return EXPRESSION_REPEATER.enabled(domain)
+
+    def allowed_to_forward(self, payload):
+        return FilterFactory.from_spec(self.configured_filter)(payload.to_json())
```
I think it'd probably be worth `memoizing` the compiled filter. (assuming the repeater object itself has a lifecycle longer than 1 request... otherwise it doesn't matter)

czue reviewed (commented):
This is awesome! I love how easy it was.

It'd be good to have some tests in the final PR.

> I believe that this can fulfill almost all of our custom API integration work

:fire: 

> The expression must be a dict expression that defines each property of the API

This is a technicality, but there's not any actual requirement of this, is there? If the API required a list, or even a single value, that would also work fine, no?

snopoke reviewed (commented):
This is very cool.

kaapstorm commented on corehq/apps/userreports/ui/fields.py:
```diff
@@ -57,4 +60,6 @@ def validate(self, value):
         if self.expected_type and not isinstance(value, self.expected_type):
-            raise forms.ValidationError(_('Expected {} but was {}'.format(self.expected_type, type(value))))
+            raise forms.ValidationError(
+                _('Expected {} but was {}'.format(self.expected_type.__name__, type(value).__name__))
```
I like that you've used `__name__` to make type names more readable.

`format` should be outside your call to `_`:
```suggestion
                _('Expected {} but was {}').format(self.expected_type.__name__, type(value).__name__)
```

kaapstorm commented on corehq/motech/repeaters/expression/repeaters.py:
```diff
@@ -0,0 +1,47 @@
+
+    @memoized
+    def payload_doc(self, repeat_record):
+        return CaseAccessors(repeat_record.domain).get_case(repeat_record.payload_id).to_json()
```
A `FormExpressionRepeater` (where the payload doc would be a form submission) might be more flexible than a `CaseExpressionRepeater`, because you can get cases from forms (using a `related_doc` expression) but I don't think you can get forms from a case.

I do agree that when we don't care about form question values, only worrying about a case is simpler.

proteusvacuum replied:
True! I think if we need a `FormExpressionRepeater` it wouldn't be hard to create.

kaapstorm reviewed (approved):
I really like this approach.

proteusvacuum committed:
Remove dict only requirement

json response can also be lists, constants, etc. not just objects

proteusvacuum committed:
Ask for permission

proteusvacuum committed:
Memoize parsed filter and expression

proteusvacuum committed:
If the value of JsonField is already a dict return it

proteusvacuum committed:
Add test

proteusvacuum commented on corehq/apps/userreports/ui/fields.py:
```diff
@@ -47,6 +47,9 @@ def prepare_value(self, value):
 
     def to_python(self, value):
+        if isinstance(value, self.expected_type):
+            return json.loads(json.dumps(value))
```
See here: https://github.com/dimagi/commcare-hq/pull/30741#discussion_r750827747

proteusvacuum commented on corehq/motech/repeaters/expression/repeaters.py:
```diff
@@ -0,0 +1,47 @@
+        return EXPRESSION_REPEATER.enabled(domain)
+
+    def allowed_to_forward(self, payload):
+        return FilterFactory.from_spec(self.configured_filter)(payload.to_json())
```
https://github.com/dimagi/commcare-hq/pull/30741/commits/484d08c3bf04a0c83fa44f67a85bfdfe6dcc5ffc

proteusvacuum commented on corehq/motech/repeaters/expression/repeater_generators.py:
```diff
@@ -0,0 +1,18 @@
+        return 'application/json'
+
+    def get_payload(self, repeat_record, payload_doc, expression):
+        parsed_expression = ExpressionFactory.from_spec(expression, FactoryContext.empty())
```
https://github.com/dimagi/commcare-hq/pull/30741/commits/484d08c3bf04a0c83fa44f67a85bfdfe6dcc5ffc

proteusvacuum committed:
Translate properly

proteusvacuum commented:
@czue 
> This is a technicality, but there's not any actual requirement of this, is there? If the API required a list, or even a single value, that would also work fine, no?

You are right! I've updated this here: https://github.com/dimagi/commcare-hq/pull/30741/commits/edfdb1acf8c90e7f3bdb0cdf04534407b805042f I keep forgetting that valid json isn't always just an object. 

@snopoke and @kaapstorm thanks for the feedback! I think I've addressed everything now. :)

czue reviewed (approved):
:ship:
